cmd/uplink/cmd/share.go

```diff
@@ -0,0 +1,166 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package cmd
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/spf13/cobra"
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/internal/fpath"
+	libuplink "storj.io/storj/lib/uplink"
+	"storj.io/storj/pkg/cfgstruct"
+	"storj.io/storj/pkg/encryption"
+	"storj.io/storj/pkg/macaroon"
+	"storj.io/storj/pkg/process"
+)
+
+var shareCfg struct {
+	DisallowReads     bool     `default:"false" help:"if true, disallow reads"`
+	DisallowWrites    bool     `default:"false" help:"if true, disallow writes"`
+	DisallowLists     bool     `default:"false" help:"if true, disallow lists"`
+	DisallowDeletes   bool     `default:"false" help:"if true, disallow deletes"`
+	Readonly          bool     `default:"false" help:"implies disallow_writes and disallow_deletes"`
+	Writeonly         bool     `default:"false" help:"implies disallow_reads and disallow_lists"`
+	NotBefore         string   `help:"disallow access before this time"`
+	NotAfter          string   `help:"disallow access after this time"`
+	AllowedPathPrefix []string `help:"whitelist of bucket path prefixes to require"`
+}
+
+func init() {
+	// sadly, we have to use addCmd so that it adds the cfg struct to the flags
+	// so that we can open projects and buckets. that pulls in so many unnecessary
+	// flags which makes figuring out the share command really hard. oh well.
+	shareCmd := addCmd(&cobra.Command{
+		Use:   "share",
+		Short: "Creates a possibly restricted api key",
+		RunE:  shareMain,
+	}, RootCmd)
+
+	cfgstruct.Bind(shareCmd.Flags(), &shareCfg)
+}
+
+const shareISO8601 = "2006-01-02T15:04:05-0700"
+
+func parseHumanDate(date string, now time.Time) (*time.Time, error) {
+	if date == "" {
+		return nil, nil
+	} else if date == "now" {
+		return &now, nil
+	} else if date[0] == '+' {
+		d, err := time.ParseDuration(date[1:])
+		t := now.Add(d)
+		return &t, errs.Wrap(err)
+	} else if date[0] == '-' {
+		d, err := time.ParseDuration(date[1:])
+		t := now.Add(-d)
+		return &t, errs.Wrap(err)
+	} else {
+		t, err := time.Parse(shareISO8601, date)
+		return &t, errs.Wrap(err)
+	}
+}
+
+// shareMain is the function executed when shareCmd is called
+func shareMain(cmd *cobra.Command, args []string) (err error) {
+	ctx := process.Ctx(cmd)
+	now := time.Now()
+
+	notBefore, err := parseHumanDate(shareCfg.NotBefore, now)
+	if err != nil {
+		return err
+	}
+	notAfter, err := parseHumanDate(shareCfg.NotAfter, now)
+	if err != nil {
+		return err
+	}
+
+	// TODO(jeff): we have to have the server side of things expecting macaroons
+	// before we can change libuplink to use macaroons because of all the tests.
+	// For now, just use the raw macaroon library.
```

Comment:
well, so, this PR does change libuplink to use macaroons, right?

Reply:
No. This TODO explains why this PR doesn't do that. That's #1945, which I'd like to get this merged, that rebased, and then it's good to review.

Reply:
oh i misread what https://github.com/storj/storj/pull/1924/files#diff-23f3da9beee4401f662ac539c8f7faefR143 was doing